src/net/client/xfr.rs

```diff
@@ -0,0 +1,1147 @@
+
+                Ok(())
+            }
+
+            State::TransferComplete => {
+                // We already finished processing an XFR response sequence. We
+                // don't expect there to be any more messages to process!.
+                Err(Error::Malformed)
+            }
+
+            State::TransferFailed => {
+                // We had to terminate processing of the XFR response sequence
+                // due to a problem with the received data, so we don't expect
+                // to be invoked again with another response message!
+                Err(Error::Terminated)
+            }
+        }
+    }
+
+    /// Check if an XFR response header is valid.
+    ///
+    /// Enforce the rules defined in 2. AXFR Messages of RFC 5936. See:
+    /// https://www.rfc-editor.org/rfc/rfc5936.html#section-2
+    ///
+    /// Takes a request as well as a response as the response is checked to
+    /// see if it is in reply to the given request.
+    ///
+    /// Returns Ok on success, Err otherwise. On success the type of XFR that
+    /// was determined is returned as well as the answer section from the XFR
+    /// response.
+    async fn check_is_xfr_answer(
+        req: &Message<Bytes>,
+        resp: &Message<Bytes>,
+    ) -> Result<XfrType, CheckError> {
+        // Check the request.
+        let req_header = req.header();
+        let req_counts = req.header_counts();
+
+        if req.is_error()
+            || req_header.qr()
+            || req_counts.qdcount() != 1
+            || req_counts.ancount() != 0
+            || req_header.opcode() != Opcode::QUERY
+        {
+            return Err(CheckError::NotValidXfrQuery);
+        }
+
+        let Some(qtype) = req.qtype() else {
+            return Err(CheckError::NotValidXfrResponse);
+        };
+
+        let xfr_type = match qtype {
+            Rtype::AXFR => XfrType::Axfr,
+            Rtype::IXFR => XfrType::Ixfr,
+            _ => return Err(CheckError::NotValidXfrResponse),
+        };
+
+        // https://datatracker.ietf.org/doc/html/rfc1995#section-3
+        // 3. Query Format
+        //   "The IXFR query packet format is the same as that of a normal DNS
+        //    query, but with the query type being IXFR and the authority
+        //    section containing the SOA record of client's version of the
+        //    zone."
+        if matches!(xfr_type, XfrType::Ixfr) && req_counts.nscount() != 1 {
+            return Err(CheckError::NotValidXfrResponse);
+        }
+
+        // Check the response.
+        let resp_header = resp.header();
+        let resp_counts = resp.header_counts();
+
+        if resp.is_error()
+            || !resp.is_answer_header(req)
+            || resp_header.opcode() != Opcode::QUERY
+            || resp_header.tc()
+            || resp_counts.ancount() == 0
+            || resp_counts.nscount() != 0
+        {
+            return Err(CheckError::NotValidXfrResponse);
+        }
```

Comment:
Duplicate checks, is_answer_header also checks that the answer section is not empty.

Reply:
I see your point. The idea was to closely replicate the complete set of checks defined in the XFR RFCs in the XFR related code and not have some checks here and some checks there. I can reduce the checks to just those that are not already done elsewhere.

Reply:
I'm resolving this because (a) `is_answer_header()` no longer exists and (b) a client cannot use `is_answer()` to verify the response because `is_answer()` disallows a valid XFR response because it requires QDCOUNT in the response to match QDCOUNT in the request, but RFC 5936 does not require subsequent AXFR responses to contain the question, so the checking makes sense to do in this code as it can't have been checked the standard way by the caller.